fjord/suggest/providers/trigger/models.py

```diff
@@ -0,0 +1,253 @@
+import re
+
+from django.db import models
+from django.utils.lru_cache import lru_cache
+
+from product_details import product_details
+
+from fjord.base.models import ModelBase
+from fjord.feedback.models import Product
+
+
+def _lang_native(locale):
+    """Look up locale code for native word
+
+    Note: This is case-sensitive. e.g. "en-US" matches English, but
+    "en-us" matches nothing.
+
+    :arg locale: locale string
+
+    :returns: language native name for that locale
+    """
+    try:
+        return product_details.languages[locale]['native']
+    except KeyError:
+        return locale
+
+
+class TriggerRuleManager(models.Manager):
+    def all_enabled(self):
+        return self.filter(is_enabled=True).order_by('sortorder')
+
+
+@lru_cache()
+def _generate_keywords_regex(keywords):
+    """Generates a Regex object to compare with
+
+    * splits keywords into parts
+    * for each part
+      * converts to unicode
+      * converts white space to '\s+'
+    * joins the parts to create a middle
+    * joins that with empty string bounds
+    * compiles it into a Regex and returns it
+
+    :arg keywords: list of words and phrases to match
+
+    :returns: regex object
+    """
+    assert isinstance(keywords, unicode)
+
+    def _escape(utext):
+        chars = '\\/.^$*+?{}()|?<>[]-'
+        for c in chars:
+            utext = utext.replace(c, u'\\' + c)
+        return u''.join(utext)
+
+    keywords = _escape(keywords)
+    parts = [
+        re.sub(u'\\s+', u'\\s+', part)
+        for part in keywords.splitlines()
+    ]
+    if len(parts) == 1:
+        middle = parts[0]
+    else:
+        middle = (
+            u'(?:' +
+            u'|'.join(parts) +
+            u')'
+        )
+    regex = u'((?:^|\\s)' + middle + u'(?:\\s|$))'
+    return re.compile(regex, re.LOCALE | re.IGNORECASE | re.UNICODE)
+
+
+class TriggerRule(ModelBase):
+    # Trigger properties
+    slug = models.SlugField(
+        unique=True,
+        help_text=(
+            u'Required: One-word unique slug for this trigger rule. Used '
+            u'in redirection urls and GA event tracking. Users will see '
+            u'this. Do not change once it is live!'
+        )
+    )
+    title = models.CharField(
+        max_length=255,
+        help_text=(
+            u'RequireD: Title of the suggestion--best to keep it a short '
```

Comment:
Plus it's the worst kind of typo that requires me to rebuild the initial migration. Boo hiss.